core/reactor.c

```diff
@@ -282,22 +216,22 @@ int _lf_do_step(void) {
 // the keepalive command-line option has not been given.
 // Otherwise, return 1.
 int next(void) {
+    // Enter the critical section and do not leave until we have
+    // determined which tag to commit to and start invoking reactions for.
+    lf_critical_section_enter();
     event_t* event = (event_t*)pqueue_peek(event_q);
     //pqueue_dump(event_q, event_q->prt);
     // If there is no next event and -keepalive has been specified
     // on the command line, then we will wait the maximum time possible.
-    // FIXME: is LLONG_MAX different from FOREVER?
-    #ifdef BIT_32
-    tag_t next_tag = { .time = LONG_MAX, .microstep = UINT_MAX};
-    #else 
-    tag_t next_tag = { .time = LLONG_MAX, .microstep = UINT_MAX};
-    #endif
+    tag_t next_tag = FOREVER_TAG_INITIALIZER;
     if (event == NULL) {
         // No event in the queue.
-        if (!keepalive_specified) { // FIXME: validator should issue a warning for unthreaded implementation
-                                    // schedule is not thread-safe
-            _lf_set_stop_tag((tag_t){.time=current_tag.time,.microstep=current_tag.microstep+1});
+        if (!keepalive_specified) {
+            _lf_set_stop_tag(
+                (tag_t){.time=current_tag.time, .microstep=current_tag.microstep+1}
+            );
         }
+        next_tag = stop_tag;
```

Comment:
```suggestion
```

Reply:
Yes, but there's also duplicate code. In addition to removing line 234, lines 243-246 can also be removed, I think.